wallet/wallet-controller/src/lib.rs

```diff
@@ -651,38 +663,55 @@ impl<T: NodeInterface + Clone + Send + Sync + 'static, W: WalletEvents> Controll
     }
 
     pub async fn generate_block(
-        &mut self,
+        &self,
         account_index: U31,
+        pool_id: Option<PoolId>,
         transactions_opt: Option<Vec<SignedTransaction>>,
     ) -> Result<Block, ControllerError<T>> {
+        let pool_id = match pool_id {
+            Some(pool_id) => pool_id,
+            None => {
+                let pools = self
+                    .wallet
+                    .get_pool_ids(account_index)
+                    .map_err(ControllerError::WalletError)?;
+                match pools.as_slice() {
+                    [] => return Err(ControllerError::NoStakingPool),
+                    [(pool_id, _)] => *pool_id,
+                    [_, _, ..] => return Err(ControllerError::MoreThanOneStakingPool),
+                }
+            }
+        };
         let pos_data = self
             .wallet
-            .get_pos_gen_block_data(account_index)
+            .get_pos_gen_block_data(account_index, pool_id)
             .map_err(ControllerError::WalletError)?;
-        let block = self
-            .rpc_client
+        self.rpc_client
             .generate_block(
                 GenerateBlockInputData::PoS(pos_data.into()),
-                transactions_opt,
+                transactions_opt.clone(),
             )
             .await
-            .map_err(ControllerError::NodeCallError)?;
-        Ok(block)
+            .map_err(ControllerError::NodeCallError)
     }
 
     pub async fn generate_blocks(
         &mut self,
         account_index: U31,
+        pool_id: Option<PoolId>,
         count: u32,
```

Comment:
Is this like the number of attempts we try to generate a block? Or number of blocks? Or is it the number of blocks assuming all attempts succeeded? 

It's very unclear what this should be called and calling it "count" makes it even worse. Please add documentation to this function explaining what this parameter does. At least, if it's the number of blocks to be generated, let's call it `block_count`.

Reply:
It's the number of blocks. I renamed it to `block_count` and added documentation.